### PR TITLE
migrate AG_NEWS to datapipes.

### PR DESCRIPTION
**Pull request by erip**
- [x] Migrate dataset in `torchtext/datasets/`
- [x] Update tests

### CODE REVIEW

**Comment by erip** [comment]
This error is a bit mysterious. 😄  I'll take a closer look at this, but local tests aren't even running as `pytest test/data/test_builtin_datasets.py` results in a segfault. Quite a mysterious situation for such a small change!

**Comment by erip** [comment]
I pulled in newest changes and it seems like CI isn't outright failing everywhere. Tests still aren't running locally, but I can debug from CI for now...

**Comment by erip** [comment]
These errors are "real" errors insomuch as these `IterDataPipe`s are actually not iterators since they don't define `__next__`. I'm not sure if this is something we want to push back to pytorch core team (as I don't think it's a torchdata issue) or if it's something we can live with. I think wrapping the dataset in an `iter` is an ugly solution... not sure. I welcome any thoughts.

**Comment by parmeet** [comment]
> I rebuilt torchtext from source and pulled down the cache and tests are now running locally. Maybe this is worth adding to `CONTRIBUTING.md` or similar?

Thanks @erip for raising this. Yes, you are right, it's not obvious that the datasets cache need to build locally (only when you look into CI that you know this). Certainly this need to be made more obvious in Contribution guidelines. At the same time, we plan to move away from caching datasets at all and rather rely on mocking. Please refer to #1493 .

**Comment by parmeet** [comment]
> These errors are "real" errors insomuch as these `IterDataPipe`s are actually not iterators since they don't define `__next__`. I'm not sure if this is something we want to push back to pytorch core team (as I don't think it's a torchdata issue) or if it's something we can live with. I think wrapping the dataset in an `iter` is an ugly solution... not sure. I welcome any thoughts.



> These errors are "real" errors insomuch as these `IterDataPipe`s are actually not iterators since they don't define `__next__`. I'm not sure if this is something we want to push back to pytorch core team (as I don't think it's a torchdata issue) or if it's something we can live with. I think wrapping the dataset in an `iter` is an ugly solution... not sure. I welcome any thoughts.

Yes, datapipes are iterable but not Iterators by design. not sure if it impose any limitation ,or if there use-cases where we want them to be Iterators? (except that this change would be BC-breaking for existing datasets, but in a good way because current datasets exhaust after one iteration ;) )

**Comment by erip** [comment]
Thanks, @parmeet. Mostly it breaks the "next test" but that can be removed if this is desirable.

**Comment by parmeet** [comment]
> Thanks, @parmeet. Mostly it breaks the "next test" but that can be removed if this is desirable.

right for testing, it's easy fix, I guess `next(iter(datapipe))` should be sufficient :)

**Comment by parmeet** [comment]
> > Thanks, @parmeet. Mostly it breaks the "next test" but that can be removed if this is desirable.
> 
> right for testing, it's easy fix, I guess `next(iter(datapipe))` should be sufficient :)

Perhaps we could completely avoid this, once we migrate the testing strategy to mocking.

**Comment by erip** [comment]
Oops, I meant [this one] (https://github.com/pytorch/text/blob/main/test/data/test_builtin_datasets.py#L93)

**Comment by parmeet** [comment]
> Oops, I meant [this one](https://github.com/pytorch/text/blob/main/test/data/test_builtin_datasets.py#L93). Adding the `next(iter(...))` here wouldn't actually be testing anything anymore, right?

ahh, I see. Got it! Yes, I think we can remove this test.

**Comment by erip** [comment]
OK, I think this one is also ready for review.

**Comment by erip** [comment]
I think this is ready for review -- it follows the cookie cutter from the Amazon Polarity PR. cc @Nayef211  @abhinavarora @parmeet.

**Review by Nayef211 — APPROVED** [review]
LGTM! For all the dataset migrations, can you add the following line to the top of the summary "Reference Issue: #1494" so we can keep track of all the PRs in issue #1494

File: torchtext/datasets/ag_news.py [hunk 1]
```diff
@@ -30,11 +31,18 @@
...
-@_wrap_split_argument(('train', 'test'))
-def AG_NEWS(root, split):
-    path = download_from_url(URL[split], root=root,
-                             path=os.path.join(root, split + ".csv"),
-                             hash_value=MD5[split],
-                             hash_type='md5')
-    return _RawTextIterableDataset(DATASET_NAME, NUM_LINES[split],
-                                   _create_data_from_csv(path))
+@_wrap_split_argument(("train", "test"))
+def AG_NEWS(root: str, split: Union[Tuple[str], str]):
```
Comment (Nayef211) [inline torchtext/datasets/ag_news.py:35]: NIT: will the type for `split` ever be `Tuple[str]`? Because as far as I understand the `@_wrap_split_argument` decorator will make `n` number of calls to this method where `n` is the number of splits in the `split` argument.

cc @parmeet

Comment (erip) [reply torchtext/datasets/ag_news.py:35]: You could ask for the train and test splits for instance, right?

Comment (parmeet) [reply torchtext/datasets/ag_news.py:35]: Yes, the split can be either tuple or just string. A user can pass something like `(train,dev,test)` which returns all the splits or just 'train' which would return only train split. The decorator is precisely taking care of this i.e to handle both the scenarios.

File: torchtext/datasets/ag_news.py [hunk 2]
```diff
@@ -30,11 +31,18 @@
...
+    if not is_module_available("torchdata"):
+        raise ModuleNotFoundError("Package `torchdata` not found. Please install following instructions at `https://github.com/pytorch/data`")
+
+    url_dp = IterableWrapper([URL[split]])
+    http_dp = HttpReader(url_dp)
+    cache_dp = http_dp.on_disk_cache(
+        filepath_fn=lambda x: os.path.join(root, split + ".csv"),
+        hash_dict={os.path.join(root, split + ".csv"): MD5[split]},
+        hash_type="md5"
+    ).end_caching(mode="w", same_filepath_fn=True)
```
Comment (parmeet) [inline torchtext/datasets/ag_news.py:45]: nit: can we update this to follow the same sequence of datapipes url->cache->download->end_caching as oppose to url->download->cache->end_caching as demonstrated in example [here](https://github.com/pytorch/data/blob/ec32ee42983760640280c937868611355e6ed281/torchdata/datapipes/iter/util/cacheholder.py#L130) and done for AmazonReviewPolarity/Full datasets? Not sure if there are any (performance) advantage by following one pattern over another. @ejguan could you please share any insights here?

Comment (ejguan) [reply torchtext/datasets/ag_news.py:45]: Yes and you should do so. The following pattern would re-download the file that violates the hash.
```suggestion
    cache_dp = url_dp.on_disk_cache(
        filepath_fn=lambda x: os.path.join(root, split + ".csv"),  # Here may need to change this function
        hash_dict={os.path.join(root, split + ".csv"): MD5[split]},
        hash_type="md5"
    )
    cache_dp = HttpReader(cache_dp)
    cache_dp = cache_dp.end_caching(mode="w", same_filepath_fn=True)
```

Comment (erip) [reply torchtext/datasets/ag_news.py:45]: Sounds good -- I'll make the fix in a moment!

File: torchtext/datasets/ag_news.py [hunk 3]
```diff
@@ -30,11 +31,18 @@
...
+        raise ModuleNotFoundError("Package `torchdata` not found. Please install following instructions at `https://github.com/pytorch/data`")
+
+    url_dp = IterableWrapper([URL[split]])
+    cache_dp = url_dp.on_disk_cache(
+        filepath_fn=lambda x: os.path.join(root, split + ".csv"),
+        hash_dict={os.path.join(root, split + ".csv"): MD5[split]},
+        hash_type="md5"
+    )
+    cache_dp = HttpReader(cache_dp)
+    cache_dp = cache_dp.end_caching(mode="w", same_filepath_fn=True)
```
Comment (Nayef211) [inline torchtext/datasets/ag_news.py:46]: NIT: can we combine these two lines like we did for [AmazonReviewFull](https://github.com/pytorch/text/pull/1499/files#diff-61c4706569928ac5352feafe9003545f246296ecc93c10e20120747f3707ccbdR52) and [AmazonReviewPolarity](https://github.com/pytorch/text/pull/1490/files#diff-daeb7a9b451f1cb20240d2fda55074b6f03d48849646c931c513050af8eff1a5R46)

We want to have as much consistency between the dataset implementations as possible

**Review by parmeet — APPROVED** [review]
LGTM!
